onnx/defs/nn/defs.cc

```diff
@@ -3684,11 +3684,16 @@ ONNX_OPERATOR_SET_SCHEMA(
               .Add("QSeqLen = Shape <start = -2, end = -1> (Q)") // q_sequence_length
               .Add("KVSeqLen = Shape <start = -2, end = -1> (K)") // kv_sequence_length
               .Const1D("NegOne", static_cast<int64_t>(-1)) // head_size, inferred from other dimensions
-              .Add("QNewShape = Concat <axis = 0> (BatchSize, QNumHeadsAttr, QSeqLen, NegOne)")
-              .Add("KVNewShape = Concat <axis = 0> (BatchSize, KVNumHeadsAttr, KVSeqLen, NegOne)")
-              .Add("QReshaped = Reshape (Q, QNewShape)")
-              .Add("KReshaped = Reshape (K, KVNewShape)")
-              .Add("VReshaped = Reshape (V, KVNewShape)")
+              // First reshape to [batch_size, seq_length, num_heads, head_size]
```

Comment:
I think the Reshape and Transpose should be added only for 3D inputs, and not for 4D inputs. Currently, we are using the num_head attributes to determine if the input is 3D or 4D. Can you make this change? Please introduce a boolean variable `is_3d_input` for this condition to make the code more readable.